kernel/posix/pthread_semaphore.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <errno.h>
+
+#ifndef _PTHREAD_SEMAPHORE_H
+#define _PTHREAD_SEMAPHORE_H
+
+int sem_destroy(sem_t *semaphore)
+{
+	if (k_sem_count_get(semaphore->sem) != 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_sem_reset(semaphore->sem);
+
+	return 0;
+}
+
+int sem_getvalue(sem_t *semaphore, int *value)
+{
+	*value = (int) k_sem_count_get(semaphore->sem);
+
+	return 0;
+}
+
+int sem_init(sem_t *semaphore, int pshared, unsigned int value)
+{
+	if ((semaphore->sem->limit) == 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	__ASSERT((pshared == 1) || (pshared == 0), "not correct pshared");
+
+	k_sem_init(semaphore->sem, value, semaphore->sem->limit);
+
+	return 0;
+}
+
+int sem_post(sem_t *semaphore)
+{
+	if (semaphore->sem == 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_sem_give(semaphore->sem);
+	return 0;
+}
+
+int sem_timedwait(sem_t *semaphore, struct timespec *abstime)
+{
+	u32_t	timeout;
+	long	msecs = 0;
+	s64_t	current_ms;
+	int	abstime_ms, temp;
+
+	__ASSERT(abstime, "abstime pointer NULL");
+
+	if ((abstime->tv_sec < 0) || (abstime->tv_nsec >= 1000000000)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	current_ms = k_uptime_get();
+	abstime_ms = _ts_to_ms(abstime);
+	msecs = abstime_ms - current_ms;
+
+	if (abstime_ms <= current_ms) {
+		timeout = 0;
+	} else {
+		timeout = _ms_to_ticks(msecs);
+	}
+
+	temp = k_sem_take(semaphore->sem, timeout);
+
+	if (temp == -EBUSY) {
+		errno = EAGAIN;
+		return -1;
+	} else if (temp != 0) {
+		errno = ETIMEDOUT;
+		return -1;
+	} else {
+		return 0;
+	}
+
+}
+
+
+int sem_trywait(sem_t *semaphore)
+{
+	int temp;
+
+	temp = k_sem_take(semaphore->sem, K_NO_WAIT);
+
+	if (temp == -EBUSY) {
+		errno = EAGAIN;
+		return -1;
+	} else if (temp != 0) {
+		errno = ETIMEDOUT;
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+int sem_wait(sem_t *semaphore)
+{
+	int temp;
+
+	temp = k_sem_take(semaphore->sem, K_FOREVER);
+
+	if (temp != 0) {
```

Comment:
Both sem_wait() and sem_trywait() can return the same error values.  By defining an auxiliary function to convert the error code from k_sem_take(), like suggested in a previous code review, this would be taken care of.